### PR TITLE
Added support to mmap IPC format

### PR DESCRIPTION
**Pull request by jorgecarleitao**
This PR adds support for memory mapping (uncompressed) IPC files into arrays, thereby avoiding data reads.

### CODE REVIEW

File: src/ffi/mmap.rs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        .iter()
+        .map(|maybe_buffer| match maybe_buffer {
+            Some(b) => *b as *const std::os::raw::c_void,
+            None => std::ptr::null(),
+        })
+        .collect::<Box<[_]>>();
+    let n_buffers = buffers.len() as i64;
+
+    let mut private_data = Box::new(PrivateData::<T> {
+        data: data.clone(),
```
Comment (ritchie46) [inline src/ffi/mmap.rs:106]: What do we clone here? E.g.` what is `T`?

File: tests/it/io/ipc/read/mmap.rs [hunk 2]
```diff
@@ -0,0 +1,40 @@
+use arrow2::error::Result;
+use arrow2::io::ipc::mmap::map_chunk;
+
+use super::super::common::read_gzip_json;
+
+#[derive(Clone)]
```
Comment (ritchie46) [inline tests/it/io/ipc/read/mmap.rs:6]: When we mmap we never get an owned `Vec`, only a `&[u8]` or a `&mut [u8]`. Do we create this `Vec` temporarily and simply never allow it to drop?

We might also store the `memmap2::Mmap` struct. This implements `AsRef<[u8]>` and will ensure the memory mapped region remains mapped until it is dropped.

File: src/ffi/mmap.rs (hunk 1 above)
Comment (jorgecarleitao) [reply src/ffi/mmap.rs:106]: `T` here is anything that implements `Clone + AsRef<[u8]>`, so, for example for `Mmap`, something like `Arc<Mmap<File>>`.

File: tests/it/io/ipc/read/mmap.rs (hunk 2 above)
Comment (jorgecarleitao) [reply tests/it/io/ipc/read/mmap.rs:6]: yeah, but `&[u8]` is technically `&'a[u8]` and our arrays are owned (they do not carry any lifetime). So, we need to control lifetime somewhere else, which is why we need an owned version of the `AsRef<[u8]>`. Moving `memmap2::Mmap` to `Arc<_>` is the most natural choice.

File: src/ffi/mmap.rs (hunk 1 above)
Comment (ritchie46) [reply src/ffi/mmap.rs:106]: Right, I was confused as `Vec<u8>` also implements those. ^^

File: tests/it/io/ipc/read/mmap.rs (hunk 2 above)
Comment (ritchie46) [reply tests/it/io/ipc/read/mmap.rs:6]: > Moving memmap2::Mmap to Arc<_> is the most natural choice.

Agreed.

**Pull request merged**